Jenkins CI support for couchdb on Power.

abdasgupta opened the pull request.

wohali commented:
Travis seems to be failing on building the JS packages, and it doesn't seem to be because of the reason for the current failure on master (i.e., log length is too long, which does need addressing).

Can you investigate please?

abdasgupta commented:
@wohali yeah.. it's failing. And still I am investing. But I am wondering where is the file create-container.sh !!

abdasgupta commented:
@wohali is there any other repo getting used in travis??

abdasgupta commented:
@wohali fixed the issue of building JS packages. But now "log length is long" error is coming. Should we rebuild??

abdasgupta commented:
@wohali , any update??

wohali commented:
Separate issue that needs resolving on master as well as this branch, I'll open a new ticket. No time to look at this until next week.

abdasgupta commented:
@wohali if anyone not currently working on it then I may work on that issue.

wohali commented:
@abdasgupta Please rebase this against `master` and see if your changes pass? Thanks.

wohali commented on build.sh:
```diff
@@ -42,11 +47,56 @@ rpms="(centos-6|centos-7)"
   # invoke as build-base <plat>
   # base images never get JavaScript, nor Erlang
-  docker build -f dockerfiles/$1 \
+  docker build -f ${DOCKERFILE} \
```
-1

wohali commented on build.sh:
```diff
@@ -77,7 +127,8 @@ build-all-js() {
 build-platform() {
-  docker build -f dockerfiles/$1 \
+  arch-check $1
+  docker build -f ${DOCKERFILE} \
```
-1, see above

wohali commented on build.sh:
```diff
@@ -297,4 +348,5 @@ EOF
     ;;
 esac
 
+rm -rf ${DOCKERFILE}
```
not necessary with the above changes.

wohali commented on dockerfiles/ubuntu-xenial:
```diff
@@ -17,7 +17,9 @@
-FROM ubuntu:xenial
+FROM BASEARCHREPO/ubuntu:xenial
+
+CROSS_BUILD_COPY bin/qemu-ARCH-static /usr/bin/
```
Again just create a new `ubuntu-xenial-ppc64le` and make your changes, no need for `sed` trickery.

abdasgupta replied:
yep, understood.

abdasgupta commented on bin/apt-dependencies.sh:
```diff
@@ -113,17 +113,29 @@ apt-get install -y vim-tiny screen
+
+    # config the CouchDB repo & install the JS packages
+    curl -H 'X-JFrog-Art-Api:AKCp5Z325Lsp8hX5UgznDy9fU1VqPAeuzsSBKsMhp1y74oHw6WfgKkPkm3FUpe2QkZ4qcEzYT'  https://na.artifactory.swg-devops.com:443/artifactory/sys-powercloud-generic-local/couch-libmozjs185-xenial-ppc64el.deb --output /tmp/couch-libmozjs185-xenial-ppc64el.deb
+    curl -H 'X-JFrog-Art-Api:AKCp5Z325Lsp8hX5UgznDy9fU1VqPAeuzsSBKsMhp1y74oHw6WfgKkPkm3FUpe2QkZ4qcEzYT'  https://na.artifactory.swg-devops.com:443/artifactory/sys-powercloud-generic-local/couch-libmozjs185-dev-xenial-ppc64el.deb --output /tmp/couch-libmozjs185-dev-xenial-ppc64el.deb
```
Actually there are no `couch-libmozjs185` packages in apache repo for ppc64le. The problem is couchdb-pkg is now throwing multiple error while building and none of them looking like Power architecture specific. One of the error is 
```
Building Apache CouchDB 2.2.0-d98e3e1
cp: cannot stat 'src/docs/build/man/apachecouchdb.1': No such file or directory
Makefile:319: recipe for target 'dist' failed
```

abdasgupta replied:
Hi @wohali, I need a clarification here. So you are suggesting that build.sh will build the JS packages and install from that. Right?

wohali replied:
Yes, step 1 is I build the JS packages (presumably on a qemu machine) and you install them by hand from local disk. Then proceed with testing the rest. Ideally do this in qemu so Travis and I can reproduce the workflow.

Once you confirm this works, I can proceed with a qemu workflow to build them and upload them myself (as it should be a CouchDB release engineer building and publishing these assets); after that, we can merge this work and add the platform.

Make sense?

abdasgupta commented on build.sh:
```diff
@@ -42,11 +47,56 @@ rpms="(centos-6|centos-7)"
+      DOCKERFILE=./$1
+      echo "Dockerfile for ppc64le updated"
+    else
+      echo "On ppc64le, only ubuntu-xenial is supported"
```
Because xenial is available for ppc64le Jenkins setup

abdasgupta commented on build.sh:
```diff
@@ -42,11 +47,56 @@ rpms="(centos-6|centos-7)"
+      BASEARCHREPO=ppc64le
+      QEMUARCH=ppc64le
+
+      cp dockerfiles/$1 .
```
Yeah. That's okay but then we have to create Dockerfiles for every alt arch. I guess that's okay.

abdasgupta commented on build.sh:
```diff
@@ -42,11 +47,56 @@ rpms="(centos-6|centos-7)"
+      echo "On ppc64le, only ubuntu-xenial is supported"
+      exit 0
+    fi
+  elif [[ "ppc64le" == `uname -m` ]]; then
```
Okay, I was future proofing. If someday in future we need to generate this image on ppc64le, then wd encounter the problem. But anyway, I can remove them for now.

abdasgupta commented:
Hi @wohali 
I already addressed your comments. But still facing some issues. First, I needed to install nodejs by extracting nodejs distribution from `https://nodejs.org/dist/` as PPA archive for nodejs is not present for ppc64le. But while I do this, dpkg is throwing error of `unmet dependencies` for nodejs while running `./build.sh couch-pkg ubuntu-xenial-ppc64le` to build fauxton. So this is a blocker now. And also travis throwing error of longer build time. Any suggestions??

wohali commented:
HI @abdasgupta ,

Can you share the error you are getting when trying to build fauxton? npm will create an error log file, just drag and drop that onto this issue and upload it as an attachment.

You may have to take your issue to the Fauxton people over at apache/couchdb-fauxton. If it's actually one of their dependencies that doesn't build right on ppc64le, you are going to have to go upstream to that package and try and get help there. That sort of build error is out of my hands. 

One thing to check, are you using a new enough version of node? 8.x should be new enough.

The long build time issue has been solved with #12. You should rebase this branch off of master, then update your `.travis.yml` file to match. We need to think about whether the `base` and `platform` targets also deserve to be added to the file, but we already have a *lot* of targets...

wohali commented:
@abdasgupta are you still working on this?

abdasgupta commented:
Hi @wohali , I want to but tackling this cascading dependency is being a problem.. Can't make fauxton built on power

wohali commented:
@abdasgupta Did you ask for help from the Fauxton team? https://github.com/apache/couchdb-fauxton/issues ?

Many of the people who work on Fauxton are also IBM employees ;)

wohali commented:
Oops, didn't mean to close.

wohali commented:
No action in a long while, closing for now.

If you're able to make headway with the Fauxton build, please open a new PR.
